hibike/virtual_device.py

```diff
@@ -0,0 +1,108 @@
+import struct
+import serial
+import hibike_message as hm
+import sys
+import random
+import time
+"""
+This program should create a virtual hibike device for testing purposes
+
+usage:
+$ socat -d -d pty,raw,echo=0 pty,raw,echo=0
+2016/09/20 21:29:03 socat[4165] N PTY is /dev/pts/26
+2016/09/20 21:29:03 socat[4165] N PTY is /dev/pts/27
+2016/09/20 21:29:03 socat[4165] N starting data transfer loop with FDs [3,3] and [5,5]
+$ python3.5 virtual_device.py -d LimitSwitch -p /dev/pts/26
+"""
+
+import argparse
+parser = argparse.ArgumentParser()
+parser.add_argument('-d', '--device', required=True, help='device type')
+parser.add_argument('-p', '--port', required=True, help='serial port')
+args = parser.parse_args()
+
+device = args.device
+port = args.port
+print(device, port)
+conn = serial.Serial(port, 115200)
+
+device_id = hm.deviceTypes[device]
+year = 1
+id = random.randint(0, 0xFFFFFFFFFFFFFFFF)
+delay = 0
+updateTime = 0
+uid = (device_id << 72) | (year << 64) | id
+
+# Here, the parameters and values to be sent in device datas are set for each device type, the list of subscribed parameters is set to empty,
+if device_id in [hm.deviceTypes["LimitSwitch"]]: 
+        subscribed_params = []
+        params_and_values = [("switch0", True), ("switch1", True), ("switch2", False), ("switch3", False)]
+if device_id in [hm.deviceTypes["ServoControl"]]:
+        subscribed_params = []
+        params_and_values = [("servo0", 2), ("enable0", True), ("servo1", 0), ("enable1", True), ("servo2", 5), ("enable2", True), ("servo3", 3), ("enable3", False)]
+if device_id in [hm.deviceTypes["Potentiometer"]]:
+        subscribed_params = []
+        params_and_values = [("pot0", 6.7), ("pot1", 5.5), ("pot2", 34.1), ("pot3", 0.15)]
+if device_id in [hm.deviceTypes["YogiBear"]]:
+        subscribed_params = []
+        params_and_values = [("duty", 20), ("forward", False)]
+        
+while (True):
+        if (updateTime != 0 and delay != 0):
+                if((time.time() - updateTime) >= (delay * 0.001)): #If the time equal to the delay has elapsed since the previous device data, send a device data with the device id and the device's subscribed params and values
+                        data = []
```

Comment:
This could be `data = [x for x in params_and_values if x[0] in subscribed_params]`